File: Analizador Semantico/Analizador/src/Interprete.java
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;

class Interprete {
    private boolean existenErrores;

    public Interprete() {
        this.existenErrores = false;
    }

    public void ejecutarArchivo(String archivo) {
        List<String> lineas = new ArrayList<>();
        try {
            File file = new File(archivo);
            Scanner scanner = new Scanner(file);
            while (scanner.hasNextLine()) {
                String linea = scanner.nextLine();
                lineas.add(linea);
            }
            scanner.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        ejecutar(lineas);
        if (existenErrores) {
            System.exit(0);
        }
    }

    public void ejecutarPrompt() {
        List<String> lineas = new ArrayList<>();
        Scanner scanner = new Scanner(System.in);
        while (true) {
            try {
                System.out.print(">>> ");
                String linea = scanner.nextLine();
                lineas.add(linea);
            } catch (Exception e) {
                break;
            }
            if (lineas.isEmpty()) {
                continue;
            }
            ejecutar(lineas);
            lineas.clear();
        }
        scanner.close();
    }

    public void error(int linea, String msj) {
        System.out.println("[line " + linea + "] | " + msj);
    }

    public void reportar(int linea, String donde, String mensaje) {
        // Implement reportar method
    }

    public void ejecutar(List<String> source) {
        Scanner scanner = new Scanner(source);
        List<String> tokens = scanner.ScanTokens();
        Parser parser = new Parser(tokens);
        parser.parse();

        Postfix postfix = new Postfix(tokens);
        List<String> postfija = postfix.convertir();

        GeneradorAST genAst = new GeneradorAST(postfija);
        Nodo programa = genAst.generarAST();
        programa.recorrer();
    }
}

public class Main {
    public static void main(String[] args) {
        Interprete interprete = new Interprete();
        init();
        if (args.length > 2) {
            System.out.println("Uso: interprete ");
            System.exit(0);
        }
        if (args.length == 2) {
            interprete.ejecutarArchivo(args[1]);
        } else {
            interprete.ejecutarPrompt();
        }
    }
}

File: AnalizadorSemantico/Interprete.py
#!/usr/local/bin/python3.11
import sys
from Scanner import Scanner
from Parser import Parser
from Postfija import Postfix
from Nodo import Nodo
from AST import GeneradorAST
from SymbolTable import init


class Interprete:
    # Clase principal del intérprete

    def __init__(self) -> None:
        self.existenErrores = False

    def ejecutarArchivo(self, archivo):
        # Método para ejecutar un archivo con código

        with open(archivo, "r") as file:
            lineas = file.readlines()

        # print(type(lineas))
        self.ejecutar(lineas)

        if self.existenErrores:
            sys.exit()

    def ejecutarPrompt(self):
        # Método para ejecutar el prompt del intérprete

        lineas = []
        while True:
            try:
                linea = input(">>> ")
                lineas.append(linea)
            except EOFError:
                break
            if not linea:
                continue
            self.ejecutar(lineas)
            lineas.clear()

    def error(linea, msj):
        # Método para reportar un error encontrado durante la ejecución del código
        print(f"[line {linea}] | {msj}")
        pass
        # self.report(self, msg, lin)

    def reportar(self, linea, donde, mensaje):
        # Método para reportar información al usuario durante la ejecución del código
        pass

    def ejecutar(self, source):
        # Método para ejecutar el código ingresado

        self.scanner = Scanner(source)
        self.tokens = self.scanner.ScanTokens()
        self.parser = Parser(self.tokens)
        self.parser.parse()
        #for i in self.tokens:
        #    print(f"{i}")
        #print("****")
        self.postfix = Postfix(self.tokens)
        self.postfija = self.postfix.convertir()
        #for i in self.postfija:
        #    print(i)
        #print("-----")
        self.genAst = GeneradorAST(self.postfija)
        self.programa = self.genAst.generarAST()
        self.programa.recorrer()


def main():
    # Función principal del intérprete
    interprete = Interprete()
    init()

    if len(sys.argv) > 2:
        print("Uso: interprete ")
        sys.exit()
    if len(sys.argv) == 2:
        interprete.ejecutarArchivo(sys.argv[1])
    else:
        interprete.ejecutarPrompt()


if __name__ == "__main__":
    main()
